bindings/profilers/heap.cc

```diff
@@ -16,44 +16,263 @@
+  }
+  Nan::Set(js_node, Nan::New<v8::String>("children").ToLocalChecked(), children);
+  v8::Local<v8::Array> allocations = Nan::New<v8::Array>(node->allocations.size());
+  for (size_t i = 0; i < node->allocations.size(); i++) {
+    v8::AllocationProfile::Allocation alloc = node->allocations[i];
+    v8::Local<v8::Object> js_alloc = Nan::New<v8::Object>();
+    Nan::Set(js_alloc, Nan::New<v8::String>("sizeBytes").ToLocalChecked(),
+             Nan::New<v8::Number>(alloc.size));
+    Nan::Set(js_alloc, Nan::New<v8::String>("count").ToLocalChecked(),
+             Nan::New<v8::Number>(alloc.count));
+    Nan::Set(allocations, i, js_alloc);
+  }
+  Nan::Set(js_node, Nan::New<v8::String>("allocations").ToLocalChecked(),
+           allocations);
+  return js_node;
+}
+
+static void dumpAllocationProfile(FILE* file, Node* node, std::string & cur_stack) {
+  auto initial_len = cur_stack.size();
+  char buf[256];
+
+  snprintf(buf, sizeof(buf), "%s%s:%s:%d", cur_stack.empty() ? "" : ";",
+           node->script_name.empty() ? "_" : node->script_name.c_str(),
+           node->name.empty() ?  "(anonymous)" : node->name.c_str(),
+           node->line_number);
+  cur_stack += buf;
+  for(auto& allocation: node->allocations) {
+    fprintf(file, "%s %u %zu\n", cur_stack.c_str(), allocation.count, allocation.count * allocation.size);
+  }
+  for(auto& child: node->children) {
+    dumpAllocationProfile(file, child.get(), cur_stack);
+  }
+  cur_stack.resize(initial_len);
+}
 
-  Nan::Set(js_node, Nan::New<String>("name").ToLocalChecked(), node->name);
-  Nan::Set(js_node, Nan::New<String>("scriptName").ToLocalChecked(),
+static void dumpAllocationProfile(FILE* file, Node* node) {
+  std::string stack;
+  dumpAllocationProfile(file, node, stack);
+}
+
+static void dumpAllocationProfileAsJSON(FILE* file, Node* node) {
+  fprintf(file, R"({"name":"%s","scriptName":"%s","scriptId":%d,"lineNumber":%d,"columnNumber":%d,"children":[)",
+          node->name.c_str(), node->script_name.c_str(), node->script_id, node->line_number, node->column_number);
+
+  bool first = true;
+  for(auto& child: node->children) {
+    if (!first) {
+      fputs(",", file);
+    } else {
+      first = false;
+    }
+    dumpAllocationProfileAsJSON(file, child.get());
+  }
+  fprintf(file, R"(],"allocations":[)");
+  first = true;
+  for(auto& allocation: node->allocations) {
+    fprintf(file, R"(%s{"sizeBytes":%zu,"count":%d})",
+            first ? "" : ",", allocation.size, allocation.count);
+    first = false;
+  }
+  fputs("]}", file);
+}
+
+static void InterruptCallback(v8::Isolate* isolate, void* data);
+static void AsyncCallback(uv_async_t* handle);
+
+static void ExportProfile(HeapProfilerState& state) {
+  char filename[L_tmpnam];
+  // tmpnam is not the recommended way to create a temporary file
+  // mkstemp would be better but is not available on windows
+  // libuv has some platform independent API but requires the
+  // event loop to be active
```

Comment:
You need the event loop to be active for the `uv_spawn` below though, right? So is there some reason _not_ to use the platform-independent thing? 🤔

Reply:
> LGTM generally. I have some minor questions/suggestions. Also wondering why you opted against using libuv for writing the files. Does that not work when near OOM?

Well, it's complicated ;)
Sorry I did not take time to explain the different approaches in the PR description.
v8 allows to register a callback with `AddNearHeapLimitCallback` that is invoked when v8 is about to OOM.
In this callback we are in a GC context, and we cannot invoke any JS synchronously.
We can extend the heap size by a given amount (the return value of the callback). If this amount is 0 or not enough to satisfy the GC, then the process will abort. If this amount is sufficient, user code will continue running, and if it continues allocating it might trigger another OOM without giving back control the event loop.
I implemented 4 different strategies to export the heap profile:
* dump it in a folded stack format on stderr: that's the poor man solution that is almost guaranteed to work
* schedule a JS callback with uv_async: this callback might never be invoked before the process goes OOM again
* schedule a callback through RequestInterrupt and call JS code from here: this seems to work, but from v8 documentation, an Isolate should not be reentered from RequestInterrupt, hence it might not be a safe thing to do
* dump the heap profile as a JSON temp file and spawn a new node process that exports the file: this seems to me the safest method and most likely to work.  In my tests, `uv_spawn` is a synchronous operation and does not require the event loop to launch the new process. The event loop needs to run to get the exit callback though, that's why I needed waitpid and some ifdefs...

Reply:
As discussed in Slack DMs, it _would_ be possible to create a new libuv loop just for this task, though that's somewhat complicated and can be left as a suggested improvement if we decide we want to actually support this feature. For now, the current approach is fine. :)